File: go/glyph/schema_evolution.go
package glyph

import (
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// ============================================================
// GLYPH Schema Evolution
// ============================================================
//
// Enables schemas to evolve safely without breaking clients. Supports:
//   - Adding new optional fields
//   - Renaming fields (with compatibility mapping)
//   - Deprecating fields
//   - Changing defaults
//   - Strict vs tolerant parsing modes
//
// Problem It Solves:
//   v1: Match{home=Arsenal away=Liverpool}
//   v2: Match{home=Arsenal away=Liverpool venue="Emirates"}
//       ↑ Parser fails on unknown field "venue"
//
// Solution:
//   Schema tracks versions and applies migrations automatically
//   v1 clients can read v2 data (missing fields get defaults)
//   v2 clients can read v1 data (new fields optional)

// EvolutionMode specifies how the schema handles version differences.
type EvolutionMode int

const (
	// ModeStrict fails on unknown fields
	ModeStrict EvolutionMode = iota
	// ModeTolerant ignores unknown fields (default)
	ModeTolerant
	// ModeMigrate auto-migrates between versions
	ModeMigrate
)

// String returns the mode name.
func (m EvolutionMode) String() string {
	switch m {
	case ModeStrict:
		return "strict"
	case ModeTolerant:
		return "tolerant"
	case ModeMigrate:
		return "migrate"
	default:
		return "unknown"
	}
}

// ============================================================
// Field Evolution Schema
// ============================================================

// EvolvingFieldType represents field types for schema evolution.
type EvolvingFieldType string

const (
	FieldTypeStr     EvolvingFieldType = "str"
	FieldTypeInt     EvolvingFieldType = "int"
	FieldTypeFloat   EvolvingFieldType = "float"
	FieldTypeBool    EvolvingFieldType = "bool"
	FieldTypeList    EvolvingFieldType = "list"
	FieldTypeDecimal EvolvingFieldType = "decimal"
)

// EvolvingField represents a field in a versioned schema.
type EvolvingField struct {
	Name         string            // Field name
	Type         EvolvingFieldType // Field type
	Required     bool              // Whether field is required
	Default      interface{}       // Default value if any
	AddedIn      string            // Version when field was added (e.g., "1.0")
	DeprecatedIn string            // Version when field was deprecated (empty if not deprecated)
	RenamedFrom  string            // Previous field name if renamed
	Validation   *regexp.Regexp    // Optional validation pattern
}

// IsAvailableIn checks if field is available in a given version.
func (f *EvolvingField) IsAvailableIn(version string) bool {
	if compareVersions(version, f.AddedIn) < 0 {
		return false // Field not added yet
	}

	if f.DeprecatedIn != "" && compareVersions(version, f.DeprecatedIn) >= 0 {
		return false // Field is deprecated
	}

	return true
}

// IsDeprecatedIn checks if field is deprecated in a given version.
func (f *EvolvingField) IsDeprecatedIn(version string) bool {
	if f.DeprecatedIn == "" {
		return false
	}
	return compareVersions(version, f.DeprecatedIn) >= 0
}

// ValidateValue validates a value against this field schema.
// Returns error message if invalid, empty string if valid.
func (f *EvolvingField) ValidateValue(value interface{}) string {
	if value == nil {
		if f.Required {
			return fmt.Sprintf("field %s is required", f.Name)
		}
		return ""
	}

	// Type checking
	switch f.Type {
	case FieldTypeStr:
		str, ok := value.(string)
		if !ok {
			return fmt.Sprintf("field %s must be string", f.Name)
		}
		if f.Validation != nil && !f.Validation.MatchString(str) {
			return fmt.Sprintf("field %s does not match pattern", f.Name)
		}
	case FieldTypeInt:
		switch value.(type) {
		case int, int32, int64:
			// OK
		default:
			return fmt.Sprintf("field %s must be int", f.Name)
		}
	case FieldTypeFloat:
		switch value.(type) {
		case float32, float64, int, int32, int64:
			// OK (allow int as float)
		default:
			return fmt.Sprintf("field %s must be float", f.Name)
		}
	case FieldTypeBool:
		if _, ok := value.(bool); !ok {
			return fmt.Sprintf("field %s must be bool", f.Name)
		}
	case FieldTypeList:
		switch value.(type) {
		case []interface{}, []string, []int, []int64, []float64:
			// OK
		default:
			return fmt.Sprintf("field %s must be list", f.Name)
		}
	}

	return ""
}

// ============================================================
// Version Schema
// ============================================================

// VersionSchema represents a schema for a specific version.
type VersionSchema struct {
	Name        string                    // Type name (e.g., "Match")
	Version     string                    // Version string (e.g., "1.0")
	Fields      map[string]*EvolvingField // Field name → schema
	Description string                    // Human-readable description
}

// NewVersionSchema creates a new version schema.
func NewVersionSchema(name, version string) *VersionSchema {
	return &VersionSchema{
		Name:    name,
		Version: version,
		Fields:  make(map[string]*EvolvingField),
	}
}

// AddField adds a field to this version.
func (vs *VersionSchema) AddField(field *EvolvingField) {
	vs.Fields[field.Name] = field
}

// GetField gets a field by name.
func (vs *VersionSchema) GetField(name string) *EvolvingField {
	return vs.Fields[name]
}

// Validate validates data against this schema version.
// Returns error message if invalid, empty string if valid.
func (vs *VersionSchema) Validate(data map[string]interface{}) string {
	// Check required fields
	for fieldName, fieldSchema := range vs.Fields {
		if fieldSchema.Required {
			if _, exists := data[fieldName]; !exists {
				return fmt.Sprintf("missing required field: %s", fieldName)
			}
		}
	}

	// Check field values
	for fieldName, value := range data {
		fieldSchema := vs.Fields[fieldName]
		if fieldSchema != nil {
			if err := fieldSchema.ValidateValue(value); err != "" {
				return err
			}
		}
	}

	return ""
}

// ============================================================
// Versioned Schema
// ============================================================

// VersionedSchema manages multiple versions of a schema.
type VersionedSchema struct {
	Name          string                    // Type name
	Versions      map[string]*VersionSchema // Version → schema
	LatestVersion string                    // Latest version string
	Mode          EvolutionMode             // How to handle version differences
}

// NewVersionedSchema creates a new versioned schema.
func NewVersionedSchema(name string) *VersionedSchema {
	return &VersionedSchema{
		Name:          name,
		Versions:      make(map[string]*VersionSchema),
		LatestVersion: "1.0",
		Mode:          ModeTolerant,
	}
}

// FieldConfig is used to configure fields when adding a version.
type FieldConfig struct {
	Type         EvolvingFieldType
	Required     bool
	Default      interface{}
	AddedIn      string
	DeprecatedIn string
	RenamedFrom  string
	Validation   string // Regex pattern
}

// AddVersion adds a version to the schema.
func (vs *VersionedSchema) AddVersion(version string, fields map[string]FieldConfig) error {
	versionSchema := NewVersionSchema(vs.Name, version)

	for fieldName, config := range fields {
		addedIn := config.AddedIn
		if addedIn == "" {
			addedIn = version
		}

		var validation *regexp.Regexp
		if config.Validation != "" {
			var err error
			validation, err = regexp.Compile(config.Validation)
			if err != nil {
				return fmt.Errorf("invalid validation pattern for field %s: %w", fieldName, err)
			}
		}

		field := &EvolvingField{
			Name:         fieldName,
			Type:         config.Type,
			Required:     config.Required,
			Default:      config.Default,
			AddedIn:      addedIn,
			DeprecatedIn: config.DeprecatedIn,
			RenamedFrom:  config.RenamedFrom,
			Validation:   validation,
		}

		versionSchema.AddField(field)
	}

	vs.Versions[version] = versionSchema
	vs.LatestVersion = vs.getLatestVersion()

	return nil
}

// GetVersion gets schema for a specific version.
func (vs *VersionedSchema) GetVersion(version string) *VersionSchema {
	return vs.Versions[version]
}

// EvolutionParseResult holds the result of parsing versioned data.
type EvolutionParseResult struct {
	Error string
	Data  map[string]interface{}
}

// Parse parses data from a specific version.
func (vs *VersionedSchema) Parse(data map[string]interface{}, fromVersion string) EvolutionParseResult {
	schema := vs.GetVersion(fromVersion)
	if schema == nil {
		return EvolutionParseResult{Error: fmt.Sprintf("unknown version: %s", fromVersion)}
	}

	// Validate
	if err := schema.Validate(data); err != "" && vs.Mode == ModeStrict {
		return EvolutionParseResult{Error: err}
	}

	result := copyMap(data)

	// Migrate to latest if needed
	if fromVersion != vs.LatestVersion {
		migrated, err := vs.migrate(data, fromVersion, vs.LatestVersion)
		if err != "" {
			return EvolutionParseResult{Error: err}
		}
		result = migrated
	}

	// Filter unknown fields in tolerant mode
	if vs.Mode == ModeTolerant {
		targetSchema := vs.GetVersion(vs.LatestVersion)
		if targetSchema != nil {
			filtered := make(map[string]interface{})
			for k, v := range result {
				if _, exists := targetSchema.Fields[k]; exists {
					filtered[k] = v
				}
			}
			result = filtered
		}
	}

	return EvolutionParseResult{Data: result}
}

// EvolutionEmitResult holds the result of emitting versioned data.
type EvolutionEmitResult struct {
	Error  string
	Header string
}

// Emit emits data for a specific version.
func (vs *VersionedSchema) Emit(data map[string]interface{}, version string) EvolutionEmitResult {
	if version == "" {
		version = vs.LatestVersion
	}

	schema := vs.GetVersion(version)
	if schema == nil {
		return EvolutionEmitResult{Error: fmt.Sprintf("unknown version: %s", version)}
	}

	// Validate
	if err := schema.Validate(data); err != "" {
		return EvolutionEmitResult{Error: err}
	}

	// Format version header
	header := fmt.Sprintf("@version %s", version)
	return EvolutionEmitResult{Header: header}
}

// migrate migrates data from one version to another.
func (vs *VersionedSchema) migrate(data map[string]interface{}, fromVersion, toVersion string) (map[string]interface{}, string) {
	currentVersion := fromVersion
	currentData := copyMap(data)

	// Get migration path
	path := vs.getMigrationPath(fromVersion, toVersion)
	if path == nil {
		return nil, fmt.Sprintf("cannot migrate from %s to %s", fromVersion, toVersion)
	}

	// Apply each migration step
	for _, nextVersion := range path {
		migrated, err := vs.migrateStep(currentData, currentVersion, nextVersion)
		if err != "" {
			return nil, err
		}
		currentData = migrated
		currentVersion = nextVersion
	}

	return currentData, ""
}

// migrateStep migrates data from one version to the next.
func (vs *VersionedSchema) migrateStep(data map[string]interface{}, fromVersion, toVersion string) (map[string]interface{}, string) {
	fromSchema := vs.GetVersion(fromVersion)
	toSchema := vs.GetVersion(toVersion)

	if fromSchema == nil || toSchema == nil {
		return nil, "invalid version"
	}

	result := copyMap(data)

	// 1. Handle field renames
	for fieldName, fieldSchema := range toSchema.Fields {
		if fieldSchema.RenamedFrom != "" {
			oldName := fieldSchema.RenamedFrom
			if oldVal, exists := result[oldName]; exists {
				if _, newExists := result[fieldName]; !newExists {
					result[fieldName] = oldVal
					delete(result, oldName)
				}
			}
		}
	}

	// 2. Handle new fields (add defaults)
	for fieldName, fieldSchema := range toSchema.Fields {
		if _, exists := result[fieldName]; !exists {
			if fieldSchema.Default != nil {
				result[fieldName] = fieldSchema.Default
			} else if !fieldSchema.Required {
				result[fieldName] = nil
			}
		}
	}

	// 3. Remove unknown fields (tolerant mode)
	if vs.Mode == ModeTolerant {
		filtered := make(map[string]interface{})
		for k, v := range result {
			if _, exists := toSchema.Fields[k]; exists {
				filtered[k] = v
			}
		}
		result = filtered
	}

	return result, ""
}

// getMigrationPath gets the migration path between versions.
func (vs *VersionedSchema) getMigrationPath(fromVersion, toVersion string) []string {
	// Get sorted versions
	versions := make([]string, 0, len(vs.Versions))
	for v := range vs.Versions {
		versions = append(versions, v)
	}
	sort.Slice(versions, func(i, j int) bool {
		return compareVersions(versions[i], versions[j]) < 0
	})

	// Find indices
	fromIdx := -1
	toIdx := -1
	for i, v := range versions {
		if v == fromVersion {
			fromIdx = i
		}
		if v == toVersion {
			toIdx = i
		}
	}

	if fromIdx == -1 || toIdx == -1 {
		return nil
	}

	if fromIdx < toIdx {
		return versions[fromIdx+1 : toIdx+1]
	} else if fromIdx > toIdx {
		// Downgrade not supported
		return nil
	}

	return []string{}
}

// getLatestVersion gets the latest version string.
func (vs *VersionedSchema) getLatestVersion() string {
	if len(vs.Versions) == 0 {
		return "1.0"
	}

	versions := make([]string, 0, len(vs.Versions))
	for v := range vs.Versions {
		versions = append(versions, v)
	}
	sort.Slice(versions, func(i, j int) bool {
		return compareVersions(versions[i], versions[j]) < 0
	})

	return versions[len(versions)-1]
}

// ChangelogEntry represents a changelog entry for a version.
type ChangelogEntry struct {
	Version          string
	Description      string
	AddedFields      []string
	DeprecatedFields []string
	RenamedFields    [][2]string // [old, new] pairs
}

// GetChangelog gets changelog of schema evolution.
func (vs *VersionedSchema) GetChangelog() []ChangelogEntry {
	// Get sorted versions
	versions := make([]string, 0, len(vs.Versions))
	for v := range vs.Versions {
		versions = append(versions, v)
	}
	sort.Slice(versions, func(i, j int) bool {
		return compareVersions(versions[i], versions[j]) < 0
	})

	changelog := make([]ChangelogEntry, 0, len(versions))

	for _, version := range versions {
		schema := vs.Versions[version]

		entry := ChangelogEntry{
			Version:          version,
			Description:      schema.Description,
			AddedFields:      []string{},
			DeprecatedFields: []string{},
			RenamedFields:    [][2]string{},
		}

		for _, field := range schema.Fields {
			if field.AddedIn == version {
				entry.AddedFields = append(entry.AddedFields, field.Name)
			}
			if field.DeprecatedIn == version {
				entry.DeprecatedFields = append(entry.DeprecatedFields, field.Name)
			}
			if field.RenamedFrom != "" {
				entry.RenamedFields = append(entry.RenamedFields, [2]string{field.RenamedFrom, field.Name})
			}
		}

		changelog = append(changelog, entry)
	}

	return changelog
}

// ============================================================
// Helper Functions
// ============================================================

// compareVersions compares two version strings.
// Returns -1 if v1 < v2, 0 if v1 == v2, 1 if v1 > v2.
func compareVersions(v1, v2 string) int {
	parts1 := parseVersionParts(v1)
	parts2 := parseVersionParts(v2)

	// Pad shorter version
	for len(parts1) < len(parts2) {
		parts1 = append(parts1, 0)
	}
	for len(parts2) < len(parts1) {
		parts2 = append(parts2, 0)
	}

	for i := range parts1 {
		if parts1[i] < parts2[i] {
			return -1
		}
		if parts1[i] > parts2[i] {
			return 1
		}
	}

	return 0
}

// parseVersionParts parses a version string into integer parts.
func parseVersionParts(version string) []int {
	parts := strings.Split(version, ".")
	result := make([]int, len(parts))

	for i, p := range parts {
		n, err := strconv.Atoi(p)
		if err != nil {
			n = 0
		}
		result[i] = n
	}

	return result
}

// copyMap creates a shallow copy of a map.
func copyMap(m map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{}, len(m))
	for k, v := range m {
		result[k] = v
	}
	return result
}

// ParseVersionHeader extracts version from a header string like "@version 2.0".
func ParseVersionHeader(text string) (string, bool) {
	text = strings.TrimSpace(text)
	if !strings.HasPrefix(text, "@version ") {
		return "", false
	}
	version := strings.TrimSpace(text[9:])
	if version == "" {
		return "", false
	}
	return version, true
}

// FormatVersionHeader creates a version header string.
func FormatVersionHeader(version string) string {
	return fmt.Sprintf("@version %s", version)
}

File: py/glyph/schema_evolution.py
"""
Schema Evolution - Safe API versioning for GLYPH

Enables schemas to evolve safely without breaking clients. Supports:
  - Adding new optional fields
  - Renaming fields (with compatibility mapping)
  - Deprecating fields
  - Changing defaults
  - Strict vs tolerant parsing modes

Problem It Solves:
  v1: Match{home away}
  v2: Match{home away venue}  ← Parser fails on unknown field
  
Solution:
  Schema tracks versions and applies migrations automatically
  v1 clients can read v2 data (missing fields get defaults)
  v2 clients can read v1 data (new fields optional)

Example:
  schema = VersionedSchema("Match")
  
  schema.add_version("1.0", {
      "home": {"type": "str", "required": True},
      "away": {"type": "str", "required": True},
  })
  
  schema.add_version("2.0", {
      "home": {"type": "str", "required": True},
      "away": {"type": "str", "required": True},
      "venue": {"type": "str", "required": False, "added_in": "2.0"},
  })
  
  # Auto-migration: v1 → v2 fills in venue=None
  # Forward compat: v1 client ignores unknown venue field
"""

from __future__ import annotations
from dataclasses import dataclass, field as dataclass_field
from typing import Dict, List, Optional, Any, Tuple
from enum import Enum
import re


class EvolutionMode(Enum):
    """Schema evolution mode."""
    STRICT = "strict"      # Fail on unknown fields
    TOLERANT = "tolerant"  # Ignore unknown fields
    MIGRATE = "migrate"    # Auto-migrate between versions


@dataclass
class FieldSchema:
    """Schema for a single field."""
    name: str
    type: str  # "str", "int", "float", "bool", "list", "decimal"
    required: bool = False
    default: Optional[Any] = None
    added_in: str = "1.0"      # Version when field was added
    deprecated_in: Optional[str] = None  # Version when field was deprecated
    renamed_from: Optional[str] = None   # Previous name if renamed
    validation: Optional[str] = None     # Validation pattern (regex)
    
    def is_available_in(self, version: str) -> bool:
        """Check if field is available in a given version."""
        if self._version_compare(version, self.added_in) < 0:
            return False  # Field not added yet
        
        if self.deprecated_in and self._version_compare(version, self.deprecated_in) >= 0:
            return False  # Field is deprecated
        
        return True
    
    def is_deprecated_in(self, version: str) -> bool:
        """Check if field is deprecated in a given version."""
        if not self.deprecated_in:
            return False
        return self._version_compare(version, self.deprecated_in) >= 0
    
    @staticmethod
    def _version_compare(v1: str, v2: str) -> int:
        """Compare two version strings. Returns: -1 if v1<v2, 0 if v1==v2, 1 if v1>v2."""
        parts1 = [int(x) for x in v1.split('.')]
        parts2 = [int(x) for x in v2.split('.')]
        
        # Pad shorter version
        while len(parts1) < len(parts2):
            parts1.append(0)
        while len(parts2) < len(parts1):
            parts2.append(0)
        
        if parts1 < parts2:
            return -1
        elif parts1 > parts2:
            return 1
        else:
            return 0
    
    def validate_value(self, value: Any) -> Optional[str]:
        """Validate a value against this field schema. Returns error message if invalid."""
        if value is None:
            if self.required:
                return f"Field {self.name} is required"
            return None
        
        # Type checking
        if self.type == "str" and not isinstance(value, str):
            return f"Field {self.name} must be string"
        elif self.type == "int" and not isinstance(value, int):
            return f"Field {self.name} must be int"
        elif self.type == "float" and not isinstance(value, (int, float)):
            return f"Field {self.name} must be float"
        elif self.type == "bool" and not isinstance(value, bool):
            return f"Field {self.name} must be bool"
        elif self.type == "list" and not isinstance(value, list):
            return f"Field {self.name} must be list"
        
        # Pattern validation
        if self.validation and isinstance(value, str):
            if not re.match(self.validation, value):
                return f"Field {self.name} does not match pattern"
        
        return None


@dataclass
class StructSchema:
    """Schema for a struct version."""
    name: str
    version: str  # e.g., "1.0", "2.0"
    fields: Dict[str, FieldSchema] = dataclass_field(default_factory=dict)
    description: str = ""
    
    def add_field(self, field: FieldSchema):
        """Add a field to this version."""
        self.fields[field.name] = field
    
    def get_field(self, name: str) -> Optional[FieldSchema]:
        """Get a field by name."""
        return self.fields.get(name)
    
    def validate(self, data: Dict[str, Any]) -> Optional[str]:
        """Validate data against this schema version."""
        # Check required fields
        for field_name, field_schema in self.fields.items():
            if field_schema.required and field_name not in data:
                return f"Missing required field: {field_name}"
        
        # Check field values
        for field_name, value in data.items():
            field_schema = self.fields.get(field_name)
            if field_schema:
                error = field_schema.validate_value(value)
                if error:
                    return error
        
        return None


@dataclass
class VersionedSchema:
    """Schema with versioning support."""
    name: str
    versions: Dict[str, StructSchema] = dataclass_field(default_factory=dict)
    latest_version: str = "1.0"
    mode: EvolutionMode = EvolutionMode.TOLERANT
    
    def add_version(self, version: str, fields: Dict[str, Dict[str, Any]]):
        """
        Add a version to the schema.
        
        Args:
            version: Version string (e.g., "1.0", "2.0")
            fields: Dict of field_name -> field_config
        """
        struct_schema = StructSchema(name=self.name, version=version)
        
        for field_name, field_config in fields.items():
            field_schema = FieldSchema(
                name=field_name,
                type=field_config.get("type", "str"),
                required=field_config.get("required", False),
                default=field_config.get("default"),
                added_in=field_config.get("added_in", version),
                deprecated_in=field_config.get("deprecated_in"),
                renamed_from=field_config.get("renamed_from"),
                validation=field_config.get("validation"),
            )
            struct_schema.add_field(field_schema)
        
        self.versions[version] = struct_schema
        self.latest_version = self._get_latest_version()
    
    def get_version(self, version: str) -> Optional[StructSchema]:
        """Get schema for a specific version."""
        return self.versions.get(version)
    
    def parse(self, data: Dict[str, Any], from_version: str) -> Tuple[Optional[str], Dict[str, Any]]:
        """
        Parse data from a specific version.

        Args:
            data: The data to parse
            from_version: Version the data is in

        Returns:
            (error_message, parsed_data)
        """
        schema = self.get_version(from_version)
        if not schema:
            return f"Unknown version: {from_version}", {}

        # Validate
        error = schema.validate(data)
        if error and self.mode == EvolutionMode.STRICT:
            return error, {}

        result = data.copy()

        # Migrate to latest if needed
        if from_version != self.latest_version:
            error, result = self._migrate(data, from_version, self.latest_version)
            if error:
                return error, {}

        # Filter unknown fields in tolerant mode
        if self.mode == EvolutionMode.TOLERANT:
            target_schema = self.get_version(self.latest_version)
            if target_schema:
                known_fields = set(target_schema.fields.keys())
                result = {k: v for k, v in result.items() if k in known_fields}

        return None, result
    
    def emit(self, data: Dict[str, Any], version: Optional[str] = None) -> Tuple[Optional[str], str]:
        """
        Emit data for a specific version.
        
        Args:
            data: The data to emit
            version: Target version (default: latest)
            
        Returns:
            (error_message, version_header)
        """
        target_version = version or self.latest_version
        
        schema = self.get_version(target_version)
        if not schema:
            return f"Unknown version: {target_version}", ""
        
        # Validate
        error = schema.validate(data)
        if error:
            return error, ""
        
        # Format version header
        header = f"@version {target_version}"
        return None, header
    
    def _migrate(self, data: Dict[str, Any], from_version: str, to_version: str) -> Tuple[Optional[str], Dict[str, Any]]:
        """Migrate data from one version to another."""
        current_version = from_version
        current_data = data.copy()
        
        # Get migration path
        path = self._get_migration_path(from_version, to_version)
        if not path:
            return f"Cannot migrate from {from_version} to {to_version}", {}
        
        # Apply each migration step
        for next_version in path:
            error, current_data = self._migrate_step(current_data, current_version, next_version)
            if error:
                return error, {}
            current_version = next_version
        
        return None, current_data
    
    def _migrate_step(self, data: Dict[str, Any], from_version: str, to_version: str) -> Tuple[Optional[str], Dict[str, Any]]:
        """Migrate data from one version to the next."""
        from_schema = self.get_version(from_version)
        to_schema = self.get_version(to_version)
        
        if not from_schema or not to_schema:
            return "Invalid version", {}
        
        result = data.copy()
        
        # 1. Handle field renames
        for field_name, field_schema in to_schema.fields.items():
            if field_schema.renamed_from:
                old_name = field_schema.renamed_from
                if old_name in result and field_name not in result:
                    result[field_name] = result.pop(old_name)
        
        # 2. Handle new fields (add defaults)
        for field_name, field_schema in to_schema.fields.items():
            if field_name not in result:
                if field_schema.default is not None:
                    result[field_name] = field_schema.default
                elif not field_schema.required:
                    result[field_name] = None
        
        # 3. Handle deprecated fields (keep for now, mark for warning)
        deprecated_fields = []
        for field_name, field_schema in to_schema.fields.items():
            if field_schema.is_deprecated_in(to_version):
                deprecated_fields.append(field_name)
        
        # 4. Remove unknown fields (tolerant mode)
        if self.mode == EvolutionMode.TOLERANT:
            known_fields = set(to_schema.fields.keys())
            result = {k: v for k, v in result.items() if k in known_fields}
        
        return None, result
    
    def _get_migration_path(self, from_version: str, to_version: str) -> Optional[List[str]]:
        """Get the migration path between versions."""
        from_parts = [int(x) for x in from_version.split('.')]
        to_parts = [int(x) for x in to_version.split('.')]
        
        # Simple linear migration (assumes versions are sequential)
        # Real implementation would handle more complex paths
        versions = sorted(self.versions.keys(), key=lambda v: [int(x) for x in v.split('.')])
        
        try:
            from_idx = versions.index(from_version)
            to_idx = versions.index(to_version)
            
            if from_idx < to_idx:
                return versions[from_idx + 1:to_idx + 1]
            elif from_idx > to_idx:
                # Downgrade not supported
                return None
            else:
                return []
        except ValueError:
            return None
    
    def _get_latest_version(self) -> str:
        """Get the latest version string."""
        if not self.versions:
            return "1.0"
        
        versions = sorted(self.versions.keys(), key=lambda v: [int(x) for x in v.split('.')])
        return versions[-1]
    
    def get_changelog(self) -> List[Dict[str, Any]]:
        """Get changelog of schema evolution."""
        changelog = []
        versions = sorted(self.versions.keys(), key=lambda v: [int(x) for x in v.split('.')])
        
        for version in versions:
            schema = self.versions[version]
            
            changes = {
                "version": version,
                "description": schema.description,
                "added_fields": [],
                "deprecated_fields": [],
                "renamed_fields": [],
            }
            
            for field in schema.fields.values():
                if field.added_in == version:
                    changes["added_fields"].append(field.name)
                if field.deprecated_in == version:
                    changes["deprecated_fields"].append(field.name)
                if field.renamed_from:
                    changes["renamed_fields"].append((field.renamed_from, field.name))
            
            changelog.append(changes)
        
        return changelog


def versioned_schema(name: str) -> VersionedSchema:
    """Create a versioned schema."""
    return VersionedSchema(name=name)


# Example usage and integration with GLYPH:
# 
# schema = versioned_schema("Match")
# 
# schema.add_version("1.0", {
#     "home": {"type": "str", "required": True},
#     "away": {"type": "str", "required": True},
# })
# 
# schema.add_version("2.0", {
#     "home": {"type": "str", "required": True},
#     "away": {"type": "str", "required": True},
#     "venue": {"type": "str", "required": False, "added_in": "2.0"},
# })
# 
# # Parsing v1 data with v2 schema
# error, data = schema.parse(
#     {"home": "Arsenal", "away": "Liverpool"},
#     from_version="1.0"
# )
# # Auto-migrates: {"home": "Arsenal", "away": "Liverpool", "venue": None}
# 
# # Emitting with version header
# error, header = schema.emit(data, version="2.0")
# # Returns: "@version 2.0"
# 
# # Get changelog
# for change in schema.get_changelog():
#     print(f"v{change['version']}: added {change['added_fields']}")
